promql/value.go

```diff
@@ -533,3 +534,67 @@ func (ssi *storageSeriesIterator) Next() chunkenc.ValueType {
 func (ssi *storageSeriesIterator) Err() error {
 	return nil
 }
+
+type fParams struct {
+	series     Series
+	constValue float64
+	isConstant bool
+	minValue   float64
+	maxValue   float64
+	hasAnyNaN  bool
+}
+
+// newFParams evaluates the expression and returns an fParams object,
+// which holds the parameter values (constant or series) along with min, max, and NaN info.
+func newFParams(ctx context.Context, ev *evaluator, expr parser.Expr) (*fParams, annotations.Annotations) {
+	if expr == nil {
+		return &fParams{}, nil
+	}
+	var constParam bool
+	if _, ok := expr.(*parser.NumberLiteral); ok {
+		constParam = true
+	}
+	val, ws := ev.eval(ctx, expr)
+	mat, ok := val.(Matrix)
+	if !ok || len(mat) == 0 {
+		return &fParams{}, ws
+	}
+	fp := &fParams{
+		series:     mat[0],
+		isConstant: constParam,
+		minValue:   math.MaxFloat64,
+		maxValue:   -math.MaxFloat64,
+	}
+
+	if constParam {
+		fp.constValue = fp.series.Floats[0].F
+		fp.minValue, fp.maxValue = fp.constValue, fp.constValue
+		fp.hasAnyNaN = math.IsNaN(fp.constValue)
+		return fp, ws
+	}
+
+	for _, v := range fp.series.Floats {
+		fp.maxValue = math.Max(fp.maxValue, v.F)
+		fp.minValue = math.Min(fp.minValue, v.F)
+		if math.IsNaN(v.F) {
+			fp.hasAnyNaN = true
+		}
+	}
+	return fp, ws
+}
+
+func (fp *fParams) Max() float64    { return fp.maxValue }
+func (fp *fParams) Min() float64    { return fp.minValue }
+func (fp *fParams) HasAnyNaN() bool { return fp.hasAnyNaN }
+
+func (fp *fParams) Next(ts int64) float64 {
```

Comment:
I think this is obscure enough to deserve a doc comment.

Currently, I don't understand the logic that checks `ts`. If it doesn't match, the method returns 0, but how does that help me? 0 could as well be the actual value – how do I know that the `ts` did not match?

Reply:
No need to check `ts` here. I'll remove it. I initially added it for consistency with `ev.nextValues`, but there it's used differently (to distinguish between float and histogram as the next sample). I'll also add doc comments for clarity.